<div style="text-align: justify; padding:5px; background-color:rgb(252, 253, 255); border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    <font color='red'>Mini Jupyter tutorial<br><br>To run each cell, click the cell and press <kbd>Run</kbd> from the menu bar. This will run any Python code or display any text within the selected cell before highlighting the next cell down. There are two types of cell: A <i>text cell</i> of type <kbd>Markdown</kbd> or <kbd>Heading</kbd> and a <i>code cell</i> of type <kbd>Code</kbd> identifiable with the <span style="font-family: courier; color:black; background-color:white;">In[ ]:</span> to the left of the cell</i>. The type of cell is also identifiable from the dropdown menu in the above menu bar to the right of <kbd>Run</kbd>. Any visual results produced by the code (text/figures) are displayed directly below that cell. Press <kbd>Run</kbd> again until you reach the end of the notebook or alternatively click <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart and Run All</kbd>. Should the Jupyter notebook crash for any reason, restart the Jupyter Kernel by clicking <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart</kbd>, and start again from the top.
        
</div>

# Tutorial 2.7: Metabolomics of Gastric Cancer: A multi-block hierarchical edge bundle

<p style="text-align: justify">
<br>
This tutorial covers the necessary steps for producing a Hierarchical Edge Bundle using multi-block data from a gastric cancer study.
</p>

<div style="text-align: justify; padding:5px; background-color:rgb(252, 253, 255); border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    <font color='red', size=4>Note: If visualisng using a JavaScript pop-up window you will need to allow pop-ups from your browser for the domain you're running from (localhost or mybinder.org).
</div> 

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    
<h2 id="1importpackagesmodules" style="text-align: justify">1. Import Packages/Modules</h2>

<p style="text-align: justify">The first code cell of this tutorial imports <a href="https://docs.python.org/3/tutorial/modules.html"><em>packages</em> and <em>modules</em></a> into the Jupyter environment. <em>Packages</em> and <em>modules</em> provide additional functions and tools beyond the in-built Python modules.
<br></p>
<br>
All the code embedded in this notebook is written using Python (<a href="http://www.python.org">python.org</a>) and JavaScript (<a href="https://www.javascript.com/">javascript.com</a>) and are built upon popular open source packages such as Networkx (<a href="https://networkx.github.io/">networkx.github.io</a>), NumPy (<a href="https://numpy.org/">numpy.org</a>), SciPy (<a href="https://www.scipy.org/">scipy.org</a>), Matplotlib (<a href="https://matplotlib.org/">matplotlib.org</a>), statsmodels (<a href="www.statsmodels.org/">statsmodels.org</a>), Scikit-learn (<a href="scikit-learn.org/">scikit-learn.org</a>), scikits.bootstrap (<a href="github.com/cgevans/scikits-bootstrap">github.com/cgevans/scikits-bootstrap</a>), Pandas (<a href="https://pandas.pydata.org/">pandas.pydata.org</a>) and D3 JavaScript (<a href="https://d3js.org/">d3js.org</a>).
    
<em>Note:</em> a tutorial focusing on the python programming language is beyond the scope of this notebook. To learn how to program in Python with Jupyter Notebook please refer to: 
<a href="https://mybinder.org/v2/gh/jakevdp/PythonDataScienceHandbook/master?filepath=notebooks%2FIndex.ipynb">Python Data Science Handbook (Jake VanderPlas, 2016)</a>.

In [1]:
import os
    
home = os.getcwd() + "/"

import numpy as np
import pandas as pd
from IPython.display import Javascript, display, IFrame
import multivis

print('All packages successfully loaded')

%load_ext autoreload
%autoreload 2

All packages successfully loaded


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

<h2 style="text-align: justify">2. Load Data and Peak Table</h2>

<p style="text-align: justify">The code cell below loads the <em>Data</em> and <em>Peak</em> tables from an Excel file using <code>loadData()</code>. When this is complete, you should see confirmation that Peak (the Peak worksheet) and Data (the Data worksheet) tables have been loaded.<br>

This dataset has previously been published in (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4716538/">Chan, A, et al. (2016)</a>) in <i>British Journal of Cancer</i> and has been put into a standardised <a href="https://en.wikipedia.org/wiki/Tidy_data">Tidy Data</a> format.
</p>

Please inspect the <a href="GastricCancer_NMR.xlsx">GastricCancer_NMR.xlsx </a>Excel file before using it in this tutorial to understand its structure. To change the dataset to be loaded into the notebook replace <code>filename = 'GastricCancer_NMR.xlsx'</code> with another file with the same <a href="https://en.wikipedia.org/wiki/Tidy_data">Tidy Data</a> format as <a href="GastricCancer_NMR.xlsx">GastricCancer_NMR.xlsx</a>, and then rerun the workflow.

</div></div>

In [2]:
file = 'GastricCancer_NMR.xlsx'

DataTable,PeakTable = multivis.utils.loadData(home + file, DataSheet='Data', PeakSheet='Peak')

Loading table: Peak
Loading table: Data
TOTAL SAMPLES: 123 TOTAL PEAKS: 149
Done!


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Data Table

Check the imported Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(DataTable)</span><br>
</div>

In [3]:
display(DataTable)

Idx    SampleID SampleType                   Class     M1       M2  \
1      2    sample_2     Sample          Gastric Cancer   43.0    525.7   
2      3    sample_3     Sample  Benign Gastric Disease  214.3  10703.2   
3      4    sample_4     Sample                 Healthy   31.6     59.7   
4      5    sample_5     Sample          Gastric Cancer   81.9    258.7   
5      6    sample_6     Sample  Benign Gastric Disease  196.9    128.2   
..   ...         ...        ...                     ...    ...      ...   
119  134  sample_134     Sample  Benign Gastric Disease  133.9    172.6   
120  135  sample_135     Sample                 Healthy    7.5    390.5   
121  137  sample_137     Sample          Gastric Cancer  405.3    510.7   
122  138  sample_138     Sample  Benign Gastric Disease   45.4    191.6   
123  139  sample_139     Sample                 Healthy   30.7     56.8   

        M3     M4     M5     M6  ...   M140    M141   M142    M143   M144  \
1    130.2    NaN  694.5  114.5  ...   84.2   357.1   16.1   455.5   29.5   
2    104.7   46.8  483.4  152.3  ...  993.5  1698.5   32.9    75.9   33.2   
3     86.4   14.0   88.6   10.3  ...   58.1    83.5   60.5   136.9   17.0   
4    315.1    8.7  243.2   18.4  ...   44.5    47.6   45.6  1441.7   35.2   
5    862.5   18.7  200.1    4.7  ...  143.8   157.2   10.4   182.1   32.6   
..     ...    ...    ...    ...  ...    ...     ...    ...     ...    ...   
119  121.5  128.4  324.6   79.7  ...   69.3   268.4    7.1    62.7   27.4   
120   67.9   38.4   58.9   26.5  ...   99.7   184.4    1.7    94.7   32.3   
121  521.9   91.9  732.1  145.7  ...  434.8    84.8  182.3   110.7  123.9   
122   41.0   18.7   40.8   32.2  ...   45.3    44.5   14.5    83.8   27.9   
123   35.9   20.9   17.4    NaN  ...   28.0    38.7    1.3   130.9   24.6   

      M145   M146   M147    M148   M149  
1     28.1   35.8  316.1   390.7  199.0  
2    802.8  967.6  154.4    31.6  195.2  
3     10.2   24.7   64.1    91.4   91.6  
4      0.1   22.8  135.0   322.3  254.3  
5    435.1  325.3  162.4   129.7  207.2  
..     ...    ...    ...     ...    ...  
119    2.8   32.5  249.6  1868.7  325.8  
120    NaN    5.8  174.2     NaN  145.5  
121    0.4   36.3   60.1   317.3  401.7  
122    0.3    0.5   47.3    47.8   46.5  
123    0.7    5.9   20.7   124.1   28.9  

[123 rows x 153 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table

Check the imported Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTable)</span><br>
</div>

In [4]:
display(PeakTable)

Idx  Name                    Label     Score      VIP1    Pvalue  \
1      1    M1        1_3-Dimethylurate  0.273432  0.921751  0.899999   
2      2    M2  1_6-Anhydro-β-D-glucose  0.386187  2.071139  0.586879   
3      3    M3     1_7-Dimethylxanthine  0.122120  2.892664  0.068365   
4      4    M4     1-Methylnicotinamide  0.781881  2.360494  0.877783   
5      5    M5           2-Aminoadipate  0.412730  2.969282  0.312104   
..   ...   ...                      ...       ...       ...       ...   
145  145  M145                    uarm1  0.226951  0.805725  0.584023   
146  146  M146                    uarm2  0.370446  2.261353  0.882148   
147  147  M147                β-Alanine  0.853027  2.679534  0.679136   
148  148  M148        π-Methylhistidine  0.116367  2.554018  0.839549   
149  149  M149        τ-Methylhistidine  0.870186  0.739459  0.445225   

        QC_RSD  
1    32.208005  
2    31.178028  
3    34.990605  
4    12.804201  
5     9.372664  
..         ...  
145  41.406985  
146  34.458172  
147  27.623517  
148  16.561921  
149   8.351801  

[149 rows x 7 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 3. Calculate the percentage of missing values in the data

</div>

In [5]:
stats_missing = multivis.utils.statistics(PeakTable, DataTable)

stats_missing.help()

Generate a table of parametric or non-parametric statistics and merges them with the Peak Table (node table).
        Initial_Parameters
            ----------
            peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
            datatable : Pandas dataframe matrix containing values for statistical analysis

        Methods
            -------
            set_params : Set parameters -
                parametric: Perform parametric statistical analysis, assuming the data is normally distributed (default: True)
                log_data: Perform a log ('natural', base 2 or base 10) on all data prior to statistical analysis (default: (False, 2))
                scale_data: Scale the data ('standard' (centers to the mean and scales to unit variance), 'minmax' (scales between 0 and 1), 'maxabs' (scales to the absolute maximum value), 'robust' (centers to the median and scales to between 25th and 75th quantile range) (default: (True, 'standard'))
        

In [6]:
params = dict({'parametric': False
              , 'log_data': (False, 2)
              , 'scale_data': (False, 'minmax')
              , 'impute_data': (False, 3)
              , 'group_column_name': 'Class'
              , 'control_group_name': 'Healthy'
              , 'group_alpha_CI': 0.05
              , 'fold_change_alpha_CI': 0.05
              , 'pca_alpha_CI': 0.05
              , 'total_missing': True
              , 'group_missing': True
              , 'pca_loadings': False
              , 'normality_test': False
              , 'group_normality_test': False
              , 'group_mean_CI': False
              , 'group_median_CI': False
              , 'mean_fold_change': False
              , 'median_fold_change': False
              , 'kruskal_wallis_test': False
              , 'levene_twoGroup': False
              , 'levene_allGroup': False
              , 'oneway_Anova_test': False
              , 'ttest_oneGroup': False
              , 'ttest_twoGroup': False
              , 'mann_whitney_u_test': False})
stats_missing.set_params(**params)

PeakTablePercentMissing = stats_missing.calculate()

In [7]:
display(PeakTablePercentMissing)

Idx  Name                    Label     Score      VIP1    Pvalue  \
0      0    M1        1_3-Dimethylurate  0.273432  0.921751  0.899999   
1      1    M2  1_6-Anhydro-β-D-glucose  0.386187  2.071139  0.586879   
2      2    M3     1_7-Dimethylxanthine  0.122120  2.892664  0.068365   
3      3    M4     1-Methylnicotinamide  0.781881  2.360494  0.877783   
4      4    M5           2-Aminoadipate  0.412730  2.969282  0.312104   
..   ...   ...                      ...       ...       ...       ...   
144  144  M145                    uarm1  0.226951  0.805725  0.584023   
145  145  M146                    uarm2  0.370446  2.261353  0.882148   
146  146  M147                β-Alanine  0.853027  2.679534  0.679136   
147  147  M148        π-Methylhistidine  0.116367  2.554018  0.839549   
148  148  M149        τ-Methylhistidine  0.870186  0.739459  0.445225   

        QC_RSD  Percent_Group_Benign Gastric Disease_Missing  \
0    32.208005                                          10.0   
1    31.178028                                           0.0   
2    34.990605                                           7.5   
3    12.804201                                           7.5   
4     9.372664                                           5.0   
..         ...                                           ...   
144  41.406985                                          10.0   
145  34.458172                                           2.5   
146  27.623517                                           0.0   
147  16.561921                                           0.0   
148   8.351801                                           0.0   

     Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  \
0                               11.627907                           12.5   
1                                2.325581                            0.0   
2                                2.325581                            7.5   
3                               16.279070                            5.0   
4                                0.000000                            0.0   
..                                    ...                            ...   
144                             13.953488                           35.0   
145                              0.000000                            7.5   
146                              2.325581                            2.5   
147                              0.000000                            5.0   
148                              0.000000                            0.0   

     Percent_Total_Missing  
0                11.382114  
1                 0.813008  
2                 5.691057  
3                 9.756098  
4                 1.626016  
..                     ...  
144              19.512195  
145               3.252033  
146               1.626016  
147               1.626016  
148               0.000000  

[149 rows x 11 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 4. Data cleaning with QC_RSD (QC relative standard deviation) and the total percentage missing values

</div>

In [8]:
# Remove any features with >=25% QC_RSD and >=10% missing values
PeakTableClean = PeakTablePercentMissing.query('QC_RSD < 25 & Percent_Total_Missing < 10')
peaklist = PeakTableClean['Name'] 
X = DataTable[peaklist]
DataTableClean = pd.merge(DataTable.T[~DataTable.T.index.isin(PeakTable['Name'])].T.reset_index(drop=True), pd.DataFrame(X, columns=peaklist).reset_index(drop=True), left_index=True, right_index=True)

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Data Table post data cleanup

Check the cleaned Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(DataTableClean)</span><br>
</div>

In [9]:
display(DataTableClean)

Idx    SampleID SampleType                   Class     M4     M5     M7  \
0      2    sample_2     Sample          Gastric Cancer    NaN  694.5   37.9   
1      3    sample_3     Sample  Benign Gastric Disease   46.8  483.4  110.1   
2      4    sample_4     Sample                 Healthy   14.0   88.6  170.3   
3      5    sample_5     Sample          Gastric Cancer    8.7  243.2  349.4   
4      6    sample_6     Sample  Benign Gastric Disease   18.7  200.1   37.3   
..   ...         ...        ...                     ...    ...    ...    ...   
118  134  sample_134     Sample  Benign Gastric Disease  128.4  324.6   38.9   
119  135  sample_135     Sample                 Healthy   38.4   58.9   42.3   
120  137  sample_137     Sample          Gastric Cancer   91.9  732.1  492.6   
121  138  sample_138     Sample  Benign Gastric Disease   18.7   40.8   46.5   
122  139  sample_139     Sample                 Healthy   20.9   17.4   12.9   

        M8     M11    M14  ...   M130   M133    M134    M137    M138   M142  \
0    125.7   490.6    NaN  ...  100.1  148.5  6674.1   938.9  6084.5   16.1   
1     85.1  2441.2   29.3  ...   42.8  129.8   787.2  1163.2   246.0   32.9   
2     23.9   140.7   62.9  ...   31.3   22.1   392.4   370.6   109.3   60.5   
3     61.1    48.7   77.8  ...    NaN   58.5   973.9   984.0  1037.4   45.6   
4    243.7   103.7   52.3  ...   75.9   76.5     NaN  2916.5  1732.1   10.4   
..     ...     ...    ...  ...    ...    ...     ...     ...     ...    ...   
118  162.0   112.7  133.9  ...   51.7   43.4  1243.5   896.6  2179.1    7.1   
119   62.3    23.2   39.0  ...   27.9   45.5   324.1   288.4    68.5    1.7   
120   95.1   194.7  166.7  ...   96.8  205.9  2731.0   683.1  3241.9  182.3   
121   25.9    50.7   26.3  ...    4.6   14.9   207.1    68.0   241.5   14.5   
122   12.0    21.7   12.9  ...   11.6   15.1   338.8   283.6   291.8    1.3   

       M143   M144    M148   M149  
0     455.5   29.5   390.7  199.0  
1      75.9   33.2    31.6  195.2  
2     136.9   17.0    91.4   91.6  
3    1441.7   35.2   322.3  254.3  
4     182.1   32.6   129.7  207.2  
..      ...    ...     ...    ...  
118    62.7   27.4  1868.7  325.8  
119    94.7   32.3     NaN  145.5  
120   110.7  123.9   317.3  401.7  
121    83.8   27.9    47.8   46.5  
122   130.9   24.6   124.1   28.9  

[123 rows x 73 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table post data cleanup

Check the cleaned Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTableClean)</span><br>
</div>

In [10]:
display(PeakTableClean)

Idx  Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
3      3    M4  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
4      4    M5        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
6      6    M7       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
7      7    M8  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
10    10   M11    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..   ...   ...                   ...       ...       ...       ...        ...   
141  141  M142                   u43  0.481533  2.794356  0.403690   7.151166   
142  142  M143                  u433  0.602162  1.008852  0.477379  24.826144   
143  143  M144                   u87  0.090389  2.602484  0.910527   6.635486   
147  147  M148     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
148  148  M149     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

     Percent_Group_Benign Gastric Disease_Missing  \
3                                             7.5   
4                                             5.0   
6                                             7.5   
7                                             0.0   
10                                           10.0   
..                                            ...   
141                                           0.0   
142                                           0.0   
143                                           0.0   
147                                           0.0   
148                                           0.0   

     Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  \
3                               16.279070                            5.0   
4                                0.000000                            0.0   
6                                0.000000                            2.5   
7                                0.000000                            0.0   
10                               4.651163                            2.5   
..                                    ...                            ...   
141                              2.325581                            0.0   
142                              4.651163                            0.0   
143                              0.000000                            0.0   
147                              0.000000                            5.0   
148                              0.000000                            0.0   

     Percent_Total_Missing  
3                 9.756098  
4                 1.626016  
6                 3.252033  
7                 0.000000  
10                5.691057  
..                     ...  
141               0.813008  
142               1.626016  
143               0.000000  
147               1.626016  
148               0.000000  

[69 rows x 11 columns]

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 5. Statistical analysis

Statistical analysis is important to identify any features or samples which may be outliers. 
It is also important to identify whether the data is normally distributed prior to any further analysis such as correlation analysis. Whether the data is normally distrubuted or not can determine the most suitable correlation function to use. For example the parametric method Pearson's correlation should be used for normally distributed data, whereas the non-parametric method Spearman's correlation is suitable for non-normally distributed data.

Statistical analysis can also provide additional univariate information for futher down-stream visualisations, such as one-way Anova p-values and PCA loadings for each feature to displayed in each of the nodes of the hierarchical edge bundle.
</div>

In [11]:
stats = multivis.utils.statistics(PeakTableClean, DataTableClean)

stats.help()

Generate a table of parametric or non-parametric statistics and merges them with the Peak Table (node table).
        Initial_Parameters
            ----------
            peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
            datatable : Pandas dataframe matrix containing values for statistical analysis

        Methods
            -------
            set_params : Set parameters -
                parametric: Perform parametric statistical analysis, assuming the data is normally distributed (default: True)
                log_data: Perform a log ('natural', base 2 or base 10) on all data prior to statistical analysis (default: (False, 2))
                scale_data: Scale the data ('standard' (centers to the mean and scales to unit variance), 'minmax' (scales between 0 and 1), 'maxabs' (scales to the absolute maximum value), 'robust' (centers to the median and scales to between 25th and 75th quantile range) (default: (True, 'standard'))
        

In [12]:
params = dict({'parametric': False
              , 'log_data': (True, 2)
              , 'scale_data': (True, 'minmax')
              , 'impute_data': (True, 3)
              , 'group_column_name': 'Class'
              , 'control_group_name': 'Healthy'
              , 'group_alpha_CI': 0.05
              , 'fold_change_alpha_CI': 0.05
              , 'pca_alpha_CI': 0.05
              , 'total_missing': False
              , 'group_missing': False
              , 'pca_loadings': True
              , 'normality_test': True
              , 'group_normality_test': False
              , 'group_mean_CI': False
              , 'group_median_CI': True
              , 'mean_fold_change': False
              , 'median_fold_change': True
              , 'kruskal_wallis_test': False
              , 'levene_twoGroup': False
              , 'levene_allGroup': False
              , 'oneway_Anova_test': False
              , 'ttest_oneGroup': False
              , 'ttest_twoGroup': False
              , 'mann_whitney_u_test': True})

stats.set_params(**params)

PeakTableStats = stats.calculate()

/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_

/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alph

/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alph

/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:860: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:276: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_cha

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table with statistical information: Shapiro-Wilk's pvalue, Mann–Whitney U-test, median concentration, median fold change and PCA loadings

Check the imported Peak Table with with statistical information by simply calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTableStats)</span><br>
</div>

In [13]:
display(PeakTableStats)

Idx  Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0     0    M4  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1     1    M5        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2     2    M7       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3     3    M8  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4     4   M11    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..  ...   ...                   ...       ...       ...       ...        ...   
64   64  M142                   u43  0.481533  2.794356  0.403690   7.151166   
65   65  M143                  u433  0.602162  1.008852  0.477379  24.826144   
66   66  M144                   u87  0.090389  2.602484  0.910527   6.635486   
67   67  M148     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
68   68  M149     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
64                                           0.0   
65                                           0.0   
66                                           0.0   
67                                           0.0   
68                                           0.0   

    Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                              16.279070                            5.0  ...   
1                               0.000000                            0.0  ...   
2                               0.000000                            2.5  ...   
3                               0.000000                            0.0  ...   
4                               4.651163                            2.5  ...   
..                                   ...                            ...  ...   
64                              2.325581                            0.0  ...   
65                              4.651163                            0.0  ...   
66                              0.000000                            0.0  ...   
67                              0.000000                            5.0  ...   
68                              0.000000                            0.0  ...   

    MannWhitneyU_BHFDR_qvalue_Benign Gastric Disease  \
0                                           0.964100   
1                                           0.390510   
2                                           0.757366   
3                                           0.964100   
4                                           0.932570   
..                                               ...   
64                                          0.192172   
65                                          0.964100   
66                                          0.389217   
67                                          0.288149   
68                                          0.592442   

    MannWhitneyU_BHFDR_qvalue_Gastric Cancer       PC1       PC2  PC1_lower  \
0                                   0.016064 -0.066985 -0.184524  -0.090591   
1                                   0.851090 -0.121359  0.022326  -0.142209   
2                                   0.193364 -0.128384  0.240563  -0.165796   
3                                   0.132866 -0.157641 -0.142328  -0.182551   
4                                   0.856002 -0.063594 -0.034657  -0.089931   
..                                       ...       ...       ...        ...   
64                                  0.204757 -0.097144  0.164271  -0.131213   
65                                  0.941582 -0.048828  0.052536  -0.076253   
66                                  0.851090 -0.029490  0.023427  -0.0572

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Determine if the data is normally distributed

If the majority of the data shows a Shapiro-Wilks pvalue > 0.05 the data is normally distributed and parametric statistical analysis is recommended, however if the pvalue is < 0.05 then the data is non-normally distributed and non-parametric analysis methods are recommended.
</div>

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Features normally distributed

In [14]:
display(len(PeakTableStats.query('Shapiro_pvalue > 0.05')))

23

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Features non-normally distributed

In [15]:
display(len(PeakTableStats.query('Shapiro_pvalue < 0.05')))

46

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Normality conclusion

The data is non-normally distributed and therefore non-parametric analysis (e.g. Kruskal-Wallis, Mann-Whitney U test, Spearman's correlation etc) is recommended.

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 6. Impute remaining missing values prior to Spearman's correlation analysis
    
Transformation can be done with a log which is commonly used for biological data types, however there are other types such as square, square root, cube root or reciprocal transformation, which may be suitable for other types of data. Scaling is highly dependent on the values found in the dataset. If the dataset contains a number of values which are very large compared to the majority of other values, then this may skew the data in favour of those values and consequently bias the results. Scaling methods such as unit variance scaling scale all the values in the dataset so that the values are all comparable. Pearson correlation, a parametric method, can later be performed where a log is necessary to normally distribute the data. However, if a non-parametric method is used, such as Spearman or Kendall's Tau, then a log is not necessary. Additionally, where a correlation is used no scaling is necessary after log transformation, as correlation analysis does not require scaling, as covariance is measured between individual values in the bivariate correlation analysis, unlike with PCA in tutorial 2.1, which performs multivariate analysis, measuring the variance across all values. However, if another similarity metric were to be used in place of correlation, such as Euclidean distance, then scaling may be a necessary step to take.
    
In this example, as Spearman's correlation is later being performed, no log or scaling is necessary, only the imputation of any remaining missing values is needed.

</div>

In [16]:
peaklist = PeakTableStats['Name']                       # Set peaklist to the metabolite names in the DataTable
X = DataTableClean[peaklist]                            # Extract X matrix from DataTableClean using peaklist
#Xlog = np.log2(X)                                      # Log transform (base-2)
#Xscale = multivis.utils.scaler(Xlog, type='standard')  # Center to a mean of 0 and scale to unit variance
X = multivis.utils.imputeData(X, k=3)                   # Impute remaining missing values using KNN impute with k=3

DataTableClean = pd.merge(DataTableClean.T[~DataTableClean.T.index.isin(PeakTable['Name'])].T.reset_index(drop=True), pd.DataFrame(X, columns=peaklist).reset_index(drop=True), left_index=True, right_index=True)

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 7. Slice up the data into blocks placed into a dictionary indexed by group/class name

Slice the data by group/class name for later identification of multi-block associations and place in a dictionary indexed by group/class name.
</div>

In [17]:
GroupByBlockPeaks, GroupByBlockData = multivis.utils.groups2blocks(PeakTableStats, DataTableClean, 'Class')

In [18]:
keys = GroupByBlockPeaks.keys()
for key in keys:
    display(GroupByBlockPeaks[key])

Idx Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0     0   A1  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1     1   A2        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2     2   A3       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3     3   A4  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4     4   A5    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..  ...  ...                   ...       ...       ...       ...        ...   
64   64  A65                   u43  0.481533  2.794356  0.403690   7.151166   
65   65  A66                  u433  0.602162  1.008852  0.477379  24.826144   
66   66  A67                   u87  0.090389  2.602484  0.910527   6.635486   
67   67  A68     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
68   68  A69     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
64                                           0.0   
65                                           0.0   
66                                           0.0   
67                                           0.0   
68                                           0.0   

    Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                              16.279070                            5.0  ...   
1                               0.000000                            0.0  ...   
2                               0.000000                            2.5  ...   
3                               0.000000                            0.0  ...   
4                               4.651163                            2.5  ...   
..                                   ...                            ...  ...   
64                              2.325581                            0.0  ...   
65                              4.651163                            0.0  ...   
66                              0.000000                            0.0  ...   
67                              0.000000                            5.0  ...   
68                              0.000000                            0.0  ...   

    MannWhitneyU_BHFDR_qvalue_Benign Gastric Disease  \
0                                           0.964100   
1                                           0.390510   
2                                           0.757366   
3                                           0.964100   
4                                           0.932570   
..                                               ...   
64                                          0.192172   
65                                          0.964100   
66                                          0.389217   
67                                          0.288149   
68                                          0.592442   

    MannWhitneyU_BHFDR_qvalue_Gastric Cancer       PC1       PC2  PC1_lower  \
0                                   0.016064 -0.066985 -0.184524  -0.090591   
1                                   0.851090 -0.121359  0.022326  -0.142209   
2                                   0.193364 -0.128384  0.240563  -0.165796   
3                                   0.132866 -0.157641 -0.142328  -0.182551   
4                                   0.856002 -0.063594 -0.034657  -0.089931   
..                                       ...       ...       ...        ...   
64                                  0.204757 -0.097144  0.164271  -0.131213   
65                                  0.941582 -0.048828  0.052536  -0.076253   
66                                  0.851090 -0.029490  0.023427  -0.057261   
67    

Idx Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0     0   B1  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1     1   B2        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2     2   B3       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3     3   B4  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4     4   B5    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..  ...  ...                   ...       ...       ...       ...        ...   
64   64  B65                   u43  0.481533  2.794356  0.403690   7.151166   
65   65  B66                  u433  0.602162  1.008852  0.477379  24.826144   
66   66  B67                   u87  0.090389  2.602484  0.910527   6.635486   
67   67  B68     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
68   68  B69     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
64                                           0.0   
65                                           0.0   
66                                           0.0   
67                                           0.0   
68                                           0.0   

    Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                              16.279070                            5.0  ...   
1                               0.000000                            0.0  ...   
2                               0.000000                            2.5  ...   
3                               0.000000                            0.0  ...   
4                               4.651163                            2.5  ...   
..                                   ...                            ...  ...   
64                              2.325581                            0.0  ...   
65                              4.651163                            0.0  ...   
66                              0.000000                            0.0  ...   
67                              0.000000                            5.0  ...   
68                              0.000000                            0.0  ...   

    MannWhitneyU_BHFDR_qvalue_Benign Gastric Disease  \
0                                           0.964100   
1                                           0.390510   
2                                           0.757366   
3                                           0.964100   
4                                           0.932570   
..                                               ...   
64                                          0.192172   
65                                          0.964100   
66                                          0.389217   
67                                          0.288149   
68                                          0.592442   

    MannWhitneyU_BHFDR_qvalue_Gastric Cancer       PC1       PC2  PC1_lower  \
0                                   0.016064 -0.066985 -0.184524  -0.090591   
1                                   0.851090 -0.121359  0.022326  -0.142209   
2                                   0.193364 -0.128384  0.240563  -0.165796   
3                                   0.132866 -0.157641 -0.142328  -0.182551   
4                                   0.856002 -0.063594 -0.034657  -0.089931   
..                                       ...       ...       ...        ...   
64                                  0.204757 -0.097144  0.164271  -0.131213   
65                                  0.941582 -0.048828  0.052536  -0.076253   
66                                  0.851090 -0.029490  0.023427  -0.057261   
67    

Idx Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0     0   C1  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1     1   C2        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2     2   C3       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3     3   C4  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4     4   C5    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..  ...  ...                   ...       ...       ...       ...        ...   
64   64  C65                   u43  0.481533  2.794356  0.403690   7.151166   
65   65  C66                  u433  0.602162  1.008852  0.477379  24.826144   
66   66  C67                   u87  0.090389  2.602484  0.910527   6.635486   
67   67  C68     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
68   68  C69     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
64                                           0.0   
65                                           0.0   
66                                           0.0   
67                                           0.0   
68                                           0.0   

    Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                              16.279070                            5.0  ...   
1                               0.000000                            0.0  ...   
2                               0.000000                            2.5  ...   
3                               0.000000                            0.0  ...   
4                               4.651163                            2.5  ...   
..                                   ...                            ...  ...   
64                              2.325581                            0.0  ...   
65                              4.651163                            0.0  ...   
66                              0.000000                            0.0  ...   
67                              0.000000                            5.0  ...   
68                              0.000000                            0.0  ...   

    MannWhitneyU_BHFDR_qvalue_Benign Gastric Disease  \
0                                           0.964100   
1                                           0.390510   
2                                           0.757366   
3                                           0.964100   
4                                           0.932570   
..                                               ...   
64                                          0.192172   
65                                          0.964100   
66                                          0.389217   
67                                          0.288149   
68                                          0.592442   

    MannWhitneyU_BHFDR_qvalue_Gastric Cancer       PC1       PC2  PC1_lower  \
0                                   0.016064 -0.066985 -0.184524  -0.090591   
1                                   0.851090 -0.121359  0.022326  -0.142209   
2                                   0.193364 -0.128384  0.240563  -0.165796   
3                                   0.132866 -0.157641 -0.142328  -0.182551   
4                                   0.856002 -0.063594 -0.034657  -0.089931   
..                                       ...       ...       ...        ...   
64                                  0.204757 -0.097144  0.164271  -0.131213   
65                                  0.941582 -0.048828  0.052536  -0.076253   
66                                  0.851090 -0.029490  0.023427  -0.057261   
67    

In [19]:
keys = GroupByBlockData.keys()
for key in keys:
    display(GroupByBlockData[key])

Idx    SampleID SampleType         A1      A2     A3     A4           A5  \
0     2    sample_2     Sample  59.566667   694.5   37.9  125.7   490.600000   
1     5    sample_5     Sample   8.700000   243.2  349.4   61.1    48.700000   
2     7    sample_7     Sample  38.866667   362.7   59.6   51.3    58.100000   
3     9    sample_9     Sample   8.400000   270.2  213.8   65.6    92.900000   
4    13   sample_13     Sample  40.900000   106.9    9.5   62.7    58.200000   
5    16   sample_16     Sample  10.400000   193.1   72.9   30.7    21.500000   
6    20   sample_20     Sample  23.700000    13.4   12.1   47.2   446.600000   
7    23   sample_23     Sample   8.000000    38.3   12.2   18.8   153.600000   
8    26   sample_26     Sample   6.000000   305.5   53.3   69.2   113.200000   
9    30   sample_30     Sample  38.800000   221.0  381.5   43.6    73.600000   
10   33   sample_33     Sample  27.600000   160.4   37.5   85.9    55.500000   
11   36   sample_36     Sample  10.300000    32.4   20.5   25.3     0.700000   
12   40   sample_40     Sample  25.333333   493.5  107.8  204.4   163.800000   
13   43   sample_43     Sample  13.800000   180.1   63.8   68.1   562.400000   
14   47   sample_47     Sample  55.900000   354.6   63.0  103.2   111.300000   
15   50   sample_50     Sample   1.600000   110.2   62.9   65.0    66.800000   
16   53   sample_53     Sample  11.200000    25.5  144.2    9.3    20.400000   
17   57   sample_57     Sample  50.500000   425.8   78.9  111.8   133.900000   
18   58   sample_58     Sample  16.900000   520.0  407.4  137.8   353.200000   
19   61   sample_61     Sample  11.000000     2.7  220.9   26.9    51.000000   
20   65   sample_65     Sample  15.700000   107.2  116.1   21.0    40.200000   
21   68   sample_68     Sample  18.233333    47.4  243.6   17.8    23.800000   
22   71   sample_71     Sample  14.133333    23.3   94.3   15.3    47.000000   
23   75   sample_75     Sample  26.300000   381.8  295.4   40.6   298.800000   
24   78   sample_78     Sample  16.600000   140.8   42.4   37.2    84.633333   
25   81   sample_81     Sample   2.100000    32.5    9.1   17.3    33.700000   
26   85   sample_85     Sample  19.500000    75.1    9.2   33.4    58.000000   
27   88   sample_88     Sample  57.400000   402.7   26.7   35.0  1688.200000   
28   92   sample_92     Sample  45.266667   321.6  279.8   23.0    36.200000   
29   95   sample_95     Sample  27.700000   514.2  465.0   55.1   205.100000   
30   98   sample_98     Sample  14.400000     1.3   31.0   78.5    23.966667   
31  102  sample_102     Sample  26.000000    53.9   10.2   48.7    57.000000   
32  105  sample_105     Sample  39.900000  2503.0  171.3   98.3   142.500000   
33  106  sample_106     Sample  36.800000   333.3   44.0   26.0   307.900000   
34  110  sample_110     Sample  56.700000   164.0   18.1   16.5   911.700000   
35  113  sample_113     Sample  17.300000    52.2   90.0   26.7    36.600000   
36  116  sample_116     Sample  35.900000   256.6   78.7   37.5    29.600000   
37  120  sample_120     Sample  22.400000   150.0   59.7   77.6   113.000000   
38  123  sample_123     Sample   1.800000   160.3   42.3   17.8    74.700000   
39  126  sample_126     Sample  32.800000    94.4   16.3   20.9   607.300000   
40  130  sample_130     Sample  24.900000    53.8   38.3   51.8   139.100000   
41  133  sample_133     Sample  53.800000    44.6   13.4   34.3   134.000000   
42  137  sample_137     Sample  91.900000   732.1  492.6   95.1   194.700000   

            A6     A7  ...          A60    A61     A62      A63     A64  \
0    60.033333  210.7  ...   100.100000  148.5  6674.1    938.9  6084.5   
1    77.800000   51.0  ...    33.866667   58.5   973.9    984.0  1037.4   
2    34.600000   58.1  ...    61.500000   85.1   695.5   1669.3  2299.3   
3    61.900000   54.2  ...    19.200000   58.1  1720.0    368.4  1317.2   
4   123.600000   83.1  ...    49.900000   42.5   766.8    443.7   508.2   
5    26.900000   91.9  ...     8.

Idx    SampleID SampleType          B1           B2          B3     B4  \
0     3    sample_3     Sample   46.800000   483.400000  110.100000   85.1   
1     6    sample_6     Sample   18.700000   200.100000   37.300000  243.7   
2    11   sample_11     Sample   47.400000   146.500000   20.600000   64.4   
3    14   sample_14     Sample   26.000000    95.500000   10.400000   26.4   
4    17   sample_17     Sample   51.100000    45.000000   58.000000   70.5   
5    21   sample_21     Sample   28.100000   353.800000  112.300000   73.3   
6    24   sample_24     Sample   31.000000  2014.700000  370.700000   80.1   
7    27   sample_27     Sample   27.800000   253.300000   16.400000   36.8   
8    31   sample_31     Sample  242.500000   170.100000   74.000000   72.2   
9    34   sample_34     Sample   43.700000   211.300000    8.000000   52.3   
10   38   sample_38     Sample   48.800000   119.600000   70.800000   81.7   
11   41   sample_41     Sample   50.100000    81.300000   30.233333   66.1   
12   44   sample_44     Sample    8.200000    30.900000  137.066667   12.9   
13   48   sample_48     Sample   18.100000    67.000000  204.300000   41.5   
14   51   sample_51     Sample   84.500000   368.400000   28.900000  121.9   
15   54   sample_54     Sample   57.100000    77.200000   19.800000   21.8   
16   59   sample_59     Sample   15.200000   229.300000   19.200000   66.5   
17   62   sample_62     Sample   39.600000    64.600000   18.100000   62.9   
18   66   sample_66     Sample   26.166667   365.000000   26.100000  117.8   
19   69   sample_69     Sample   19.900000   273.300000   22.200000   55.8   
20   72   sample_72     Sample    0.600000   138.500000   28.000000   54.6   
21   76   sample_76     Sample   90.500000    68.800000   21.900000   44.1   
22   79   sample_79     Sample  173.900000   366.400000  224.400000  129.8   
23   83   sample_83     Sample   79.900000    67.200000    4.600000   49.6   
24   86   sample_86     Sample   37.700000   543.800000  169.900000   73.4   
25   89   sample_89     Sample   40.400000   254.800000   35.000000   73.9   
26   93   sample_93     Sample   56.800000   207.666667   17.500000   22.0   
27   96   sample_96     Sample   45.400000   218.000000   38.600000   61.4   
28   99   sample_99     Sample   18.800000    35.000000   12.700000   54.5   
29  103  sample_103     Sample   31.066667    65.200000   10.200000   49.9   
30  107  sample_107     Sample  100.500000   215.800000   32.200000   67.4   
31  111  sample_111     Sample  203.600000  2143.100000   83.300000  525.0   
32  114  sample_114     Sample   51.300000   165.600000   51.100000   59.0   
33  117  sample_117     Sample   32.700000   304.700000   30.300000   76.0   
34  121  sample_121     Sample    0.100000   117.800000   50.800000   37.9   
35  124  sample_124     Sample   39.700000   131.900000   22.500000   53.6   
36  128  sample_128     Sample   73.900000    67.000000   16.700000   22.4   
37  131  sample_131     Sample   38.300000   167.700000   25.800000   90.6   
38  134  sample_134     Sample  128.400000   324.600000   38.900000  162.0   
39  138  sample_138     Sample   18.700000    40.800000   46.500000   25.9   

             B5          B6     B7  ...         B60     B61          B62  \
0   2441.200000   29.300000   45.4  ...   42.800000   129.8   787.200000   
1    103.700000   52.300000   76.8  ...   75.900000    76.5   770.766667   
2     38.500000   42.000000   62.2  ...   11.300000    31.1  1625.100000   
3    215.000000   19.100000   39.3  ...   80.800000    40.5   254.800000   
4    161.900000  159.200000   69.7  ...   16.600000  1039.6  1646.200000   
5    618.466667   56.700000   94.7  ...  168.400000    82.7  1285.000000   
6     36.633333  123.700000   60.1  ...  104.800000   270.4  5438.400000   
7     25.800000   29.700000   79.0  ...    9.700000    72.2   397.900000   
8    413.000000   62.900000   73.5  ...   11.900000   134.8   925.600000   
9     36.800000   61.100000   35.7  ...   79.

Idx    SampleID SampleType          C1     C2          C3     C4  \
0     4    sample_4     Sample   14.000000   88.6  170.300000   23.9   
1     8    sample_8     Sample   18.200000   72.5   15.300000   37.1   
2    12   sample_12     Sample   45.000000   62.6   42.400000   68.0   
3    15   sample_15     Sample   70.600000   65.4   26.200000   81.2   
4    18   sample_18     Sample   13.400000   51.2   23.600000   27.9   
5    22   sample_22     Sample   51.400000  228.6   25.800000  207.4   
6    25   sample_25     Sample   91.800000  551.7  173.400000  112.9   
7    29   sample_29     Sample   58.500000  187.8  166.300000   53.6   
8    32   sample_32     Sample   58.300000  227.2   29.700000  118.0   
9    35   sample_35     Sample   70.300000   43.7   28.300000   99.0   
10   39   sample_39     Sample  109.800000  267.5  316.600000   81.6   
11   42   sample_42     Sample    8.700000   25.8   10.500000    9.3   
12   45   sample_45     Sample   72.700000   67.3   16.300000   50.4   
13   49   sample_49     Sample    7.500000   22.3   13.200000   40.0   
14   52   sample_52     Sample   11.633333   31.3    9.400000   19.5   
15   56   sample_56     Sample    9.533333   45.2   25.100000   22.3   
16   60   sample_60     Sample   49.400000  237.7   28.000000  121.1   
17   63   sample_63     Sample   71.400000  452.3   45.700000  161.0   
18   67   sample_67     Sample  109.400000  510.8  276.100000  126.2   
19   70   sample_70     Sample   45.700000   65.0   17.100000   49.3   
20   74   sample_74     Sample  136.300000  236.2   47.200000  266.6   
21   77   sample_77     Sample  141.800000  360.5   38.800000   57.7   
22   80   sample_80     Sample   16.000000   48.2   13.200000   40.7   
23   84   sample_84     Sample   25.400000  125.1   39.300000   41.2   
24   87   sample_87     Sample   79.800000  252.4   18.000000   71.4   
25   90   sample_90     Sample  144.800000  460.0   54.300000  126.1   
26   94   sample_94     Sample   46.200000   92.7    8.600000   47.7   
27   97   sample_97     Sample   41.400000  175.3   10.800000   49.7   
28  101  sample_101     Sample   52.500000  170.1   45.300000  118.4   
29  104  sample_104     Sample   24.800000   35.2   12.700000   17.3   
30  108  sample_108     Sample    7.800000  143.5   12.100000   26.7   
31  112  sample_112     Sample   39.900000   54.0  152.500000  281.1   
32  115  sample_115     Sample   15.500000   72.8   16.600000   51.2   
33  119  sample_119     Sample   49.000000   77.5   19.033333  114.2   
34  122  sample_122     Sample   19.400000   79.2   13.800000   66.2   
35  125  sample_125     Sample   86.400000  153.5   31.400000   89.2   
36  129  sample_129     Sample    7.300000  822.9   52.800000   99.8   
37  132  sample_132     Sample    6.400000   56.7   23.600000   21.5   
38  135  sample_135     Sample   38.400000   58.9   42.300000   62.3   
39  139  sample_139     Sample   20.900000   17.4   12.900000   12.0   

             C5     C6     C7  ...         C60     C61          C62  \
0    140.700000   62.9   38.3  ...   31.300000    22.1   392.400000   
1     54.100000   30.3   19.2  ...   57.000000    22.3   887.700000   
2    100.700000   45.5   60.8  ...   12.900000    24.3   763.700000   
3     73.700000   95.8   48.8  ...    2.500000    53.8  1203.700000   
4     58.200000   25.4   24.8  ...    5.300000    20.9   153.700000   
5    482.300000  295.8  157.0  ...  664.900000   185.2  1966.400000   
6    105.100000  132.4  212.3  ...   33.200000   149.3  1064.800000   
7      5.700000   92.2   70.7  ...   98.600000    61.8   874.700000   
8     77.300000   98.6  108.0  ...    1.500000    65.2   722.900000   
9    295.600000  105.2  109.7  ...   34.400000   583.1   903.400000   
10    59.400000   39.1   55.0  ...   12.800000   115.8    64.600000   
11    20.600000   13.3    7.9  ...    1.000000     4.1    83.600000   
12   174.000000  114.1   44.3  ...   12.000000   933.4   113.500000   
13    29.000000  102.6   45.9  ...  157.133333    25.2   

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 8. Merge all the data
    
Merge all the data from each group/block and consolidate any statistical results generated from the multivis.utils.statistics package in relation to each block for later identification of multi-block associations and visualisation.
    
The multi-block data can be prepared using multivis.utils.groups2blocks given a single dataset separated by a group/class column and then merged together on a common index. This is suitable for when comparing differences between groups of samples which do not have a common sample ID. If the dimensions of the index are different for each group, any sample outliers can be identified through exploratory analysis (e.g. PCA) and removed until the dimensions of each group are the same, or if there are no outliers simply merge the data on the common index values and any samples outside the common index will be removed automatically. When the sample IDs are common between each block of data, such as when analysing time-series data comparing the same individuals over time or the same samples across different 'Omics platforms through a systems biology approach (e.g. transcriptomics, proteomics, metabolomics), the multi-block data from each of the "omics" blocks can be placed into a dictionary indexed by the source name. However, the multi-omics approach does require prior modeling with methods such as multi-block variable influence on orthogonal projections (MB-VIOP) and OnPLS or methods such as <a href="http://mixomics.org/">mixOmics</a>. In such a case merging on a common sample ID should be used instead of merging on the index. A good example of this can be seen in this study on asthma by <a href="https://pubs.acs.org/doi/10.1021/acs.analchem.8b03205">Reinke, S et. al. (2018)</a>, where the same samples are measured across multiple 'omics platforms and then log-transformed, scaled and modelled using MB-VIOP and OnPLS.
</div>

In [20]:
MultiBlockPeaks, MultiBlockData = multivis.utils.mergeBlocks(GroupByBlockPeaks, GroupByBlockData, 'Index')

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Multi-block Peak Table

Check the Multi-block Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(MultiBlockPeaks)</span><br>
</div>

In [21]:
display(MultiBlockPeaks)

Idx Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0      0   A1  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1      1   A2        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2      2   A3       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3      3   A4  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4      4   A5    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..   ...  ...                   ...       ...       ...       ...        ...   
202  202  C65                   u43  0.481533  2.794356  0.403690   7.151166   
203  203  C66                  u433  0.602162  1.008852  0.477379  24.826144   
204  204  C67                   u87  0.090389  2.602484  0.910527   6.635486   
205  205  C68     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
206  206  C69     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

     Percent_Group_Benign Gastric Disease_Missing  \
0                                             7.5   
1                                             5.0   
2                                             7.5   
3                                             0.0   
4                                            10.0   
..                                            ...   
202                                           0.0   
203                                           0.0   
204                                           0.0   
205                                           0.0   
206                                           0.0   

     Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                               16.279070                            5.0  ...   
1                                0.000000                            0.0  ...   
2                                0.000000                            2.5  ...   
3                                0.000000                            0.0  ...   
4                                4.651163                            2.5  ...   
..                                    ...                            ...  ...   
202                              2.325581                            0.0  ...   
203                              4.651163                            0.0  ...   
204                              0.000000                            0.0  ...   
205                              0.000000                            5.0  ...   
206                              0.000000                            0.0  ...   

     MedianFoldChange_CI_lower  MedianFoldChange_CI_upper  \
0                     0.806215                   1.006288   
1                     0.981900                   1.325797   
2                     1.189610                   2.049094   
3                     0.613169                   1.103635   
4                     0.856456                   1.067833   
..                         ...                        ...   
202                        NaN                        NaN   
203                        NaN                        NaN   
204                        NaN                        NaN   
205                        NaN                        NaN   
206                        NaN                        NaN   

     MedianFoldChange_sig  Group_median  Group_median_CI_lower  \
0                    True      0.694372               0.637680   
1                    True      0.636703               0.583049   
2                    True      0.559626               0.453478   
3                    True      0.383064               0.296091   
4                    True      0.622464               0.576839   
..                    ...           ...                    ...   
202                 False      0.516971               0.456014   
203                 False      0.655378               0.606711   
204                 False      0.107615               0.099250   
205                 False      0.574362  

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Multi-block Data Table

Check the Multi-block Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(MultiBlockData)</span><br>
</div>

In [22]:
display(MultiBlockData)

Idx SampleType         A1      A2     A3     A4           A5          A6  \
0     2     Sample  59.566667   694.5   37.9  125.7   490.600000   60.033333   
1     5     Sample   8.700000   243.2  349.4   61.1    48.700000   77.800000   
2     7     Sample  38.866667   362.7   59.6   51.3    58.100000   34.600000   
3     9     Sample   8.400000   270.2  213.8   65.6    92.900000   61.900000   
4    13     Sample  40.900000   106.9    9.5   62.7    58.200000  123.600000   
5    16     Sample  10.400000   193.1   72.9   30.7    21.500000   26.900000   
6    20     Sample  23.700000    13.4   12.1   47.2   446.600000   64.900000   
7    23     Sample   8.000000    38.3   12.2   18.8   153.600000   50.300000   
8    26     Sample   6.000000   305.5   53.3   69.2   113.200000   50.900000   
9    30     Sample  38.800000   221.0  381.5   43.6    73.600000   81.500000   
10   33     Sample  27.600000   160.4   37.5   85.9    55.500000   29.100000   
11   36     Sample  10.300000    32.4   20.5   25.3     0.700000   73.900000   
12   40     Sample  25.333333   493.5  107.8  204.4   163.800000   56.100000   
13   43     Sample  13.800000   180.1   63.8   68.1   562.400000  241.500000   
14   47     Sample  55.900000   354.6   63.0  103.2   111.300000  176.700000   
15   50     Sample   1.600000   110.2   62.9   65.0    66.800000   67.300000   
16   53     Sample  11.200000    25.5  144.2    9.3    20.400000   22.500000   
17   57     Sample  50.500000   425.8   78.9  111.8   133.900000   44.800000   
18   58     Sample  16.900000   520.0  407.4  137.8   353.200000  196.800000   
19   61     Sample  11.000000     2.7  220.9   26.9    51.000000   28.300000   
20   65     Sample  15.700000   107.2  116.1   21.0    40.200000   23.800000   
21   68     Sample  18.233333    47.4  243.6   17.8    23.800000   37.000000   
22   71     Sample  14.133333    23.3   94.3   15.3    47.000000   26.700000   
23   75     Sample  26.300000   381.8  295.4   40.6   298.800000   45.300000   
24   78     Sample  16.600000   140.8   42.4   37.2    84.633333   19.600000   
25   81     Sample   2.100000    32.5    9.1   17.3    33.700000   18.600000   
26   85     Sample  19.500000    75.1    9.2   33.4    58.000000   34.600000   
27   88     Sample  57.400000   402.7   26.7   35.0  1688.200000   50.600000   
28   92     Sample  45.266667   321.6  279.8   23.0    36.200000   35.200000   
29   95     Sample  27.700000   514.2  465.0   55.1   205.100000   34.500000   
30   98     Sample  14.400000     1.3   31.0   78.5    23.966667   50.400000   
31  102     Sample  26.000000    53.9   10.2   48.7    57.000000   43.300000   
32  105     Sample  39.900000  2503.0  171.3   98.3   142.500000   83.700000   
33  106     Sample  36.800000   333.3   44.0   26.0   307.900000   39.000000   
34  110     Sample  56.700000   164.0   18.1   16.5   911.700000    0.500000   
35  113     Sample  17.300000    52.2   90.0   26.7    36.600000   21.500000   
36  116     Sample  35.900000   256.6   78.7   37.5    29.600000   44.100000   
37  120     Sample  22.400000   150.0   59.7   77.6   113.000000   94.300000   
38  123     Sample   1.800000   160.3   42.3   17.8    74.700000   17.000000   
39  126     Sample  32.800000    94.4   16.3   20.9   607.300000   40.600000   

       A7          A8  ...         C60     C61          C62           C63  \
0   210.7  150.300000  ...   31.300000    22.1   392.400000    370.600000   
1    51.0   71.300000  ...   57.000000    22.3   887.700000    260.800000   
2    58.1  277.200000  ...   12.900000    24.3   763.700000    617.400000   
3    54.2   85.000000  ...    2.500000    53.8  1203.700000     99.200000   
4    83.1   62.500000  ...    5.300000    20.9   153.700000    180.400000   
5    91.9   54.500000  ...  664.900000   185.2  1966.400000    821.100000   
6    67.4   52.700000  ...   33.200000   149.3  1064.800000    512.700000   
7    18.5   21.366667  ...   98.600000    61.8   874.700000    108.400000   
8   109.1  665.900000  ...    

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 9. Correlation analysis

Correlation is a form of similarity and measures the strength of the linear relationship between two variables. Pearson's correlation, a form of parametric correlation analysis, is described mathematically by dividing the joint variability or covariance of two variables by the product of their standard deviations (see Eq1). Other forms of correlation measure the monotonic relationships and are non-parametric, such as Spearman’s rank correlation and Kendall Tau's correlation. The following correlation analysis, allows for Pearson, Spearman or Kendall Tau's correlation analysis.

\begin{equation*}
r = \frac{Cov(X,Y)}{SD(X).SD(Y)}
\end{equation*}
<center>Eq1: Pearson’s correlation coefficient</center>

</div>

In [23]:
correlationType = "spearman" #"pearson"; "kendalltau"

X = MultiBlockData[MultiBlockPeaks['Name']]

MultiBlockScores,MultiBlockPvalues = multivis.utils.corrAnalysis(X, correlationType)

100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 207/207 [00:11<00:00, 17.35it/s]


<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 10.  Generate Edges

The similarities/distances are filtered and put in a dataframe of edges, where nodes represent features (metabolites) and edges represent similarity/distance scores (e.g. correlation coefficients), with included node names, labels and other attributes such as correlation coefficient pvalues and other statistical information.

</div>

In [24]:
networkEdges = multivis.Edge(peaktable=MultiBlockPeaks, datatable=MultiBlockScores, pvalues=MultiBlockPvalues)

networkEdges.help()

Builds nodes and edges and is the base class for the Network class.

        Initial_Parameters
        ----------
        peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
        datatable : Pandas dataframe matrix containing scores
        pvalues : Pandas dataframe matrix containing score/similarity pvalues (if available, otherwise set to None)

        Methods
        -------
        set_params : Set parameters
            filter_type: The value type to filter the data on (default: 'pvalue')
            hard_threshold: Value to filter the data on (default: 0.005)
            withinBlocks: Include scores within blocks if building multi-block network (default: False)
            sign: The sign of the score/similarity to filter on ('pos', 'neg' or 'both') (default: 'both')

        help : Print this help text

        build : Builds the nodes and edges.
        getNodes : Returns a Pandas dataframe of all nodes.
        getEdges : Returns a Pandas da

In [25]:
params = dict({'filter_type': 'pvalue'              #The filer type to use for the similarities matrix ('Pvalue' or 'Score')              
                    , 'hard_threshold': 0.05        #The hard threshold to apply to the similarities matrix
                    , 'withinBlocks': True          #Include scores within blocks if building multi-block network
                    , 'sign': "both"})              #The sign of the similarities ('pos', 'neg' or 'both')

networkEdges.set_params(**params)

networkEdges.build()

edges = networkEdges.getEdges()
nodes = networkEdges.getNodes()

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display node data used in the Hierarchical Edge Bundle

Check the node data simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(nodes)</span><br>
</div>

In [26]:
display(nodes)

Idx Name                 Label                Score                VIP1  \
0      0   A1  1-Methylnicotinamide   0.7818809592495104  2.3604940845644222   
1      1   A2        2-Aminoadipate  0.41272975250933697   2.969282495538328   
2      2   A3       2-Furoylglycine   0.6216059593217798  2.9102253505330196   
3      3   A4  2-Hydroxyisobutyrate  0.33991383587817403  0.6368605295964379   
4      4   A5    3-Aminoisobutyrate  0.05261303702988451  1.2322610580806348   
..   ...  ...                   ...                  ...                 ...   
202  202  C65                   u43  0.48153258586917613  2.7943561790255034   
203  203  C66                  u433   0.6021621387517222   1.008852452462372   
204  204  C67                   u87   0.0903890065459535  2.6024844722124936   
205  205  C68     π-Methylhistidine  0.11636744927699838  2.5540178928384805   
206  206  C69     τ-Methylhistidine   0.8701858668724608  0.7394589178223214   

                  Pvalue            QC_RSD  \
0     0.8777833638996088  12.8042008738419   
1     0.3121038626986239  9.37266355735233   
2     0.4223951783297637  5.04915562556207   
3    0.04047086649009335  5.13234006385694   
4     0.9112069804740864  15.4761649136304   
..                   ...               ...   
202  0.40369030143761075  7.15116605857213   
203  0.47737879764439495  24.8261437338676   
204   0.9105266138515601  6.63548612638398   
205   0.8395491475231609  16.5619208190305   
206  0.44522519478107525   8.3518014361243   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
202                                          0.0   
203                                          0.0   
204                                          0.0   
205                                          0.0   
206                                          0.0   

    Percent_Group_Gastric Cancer_Missing Percent_Group_Healthy_Missing  ...  \
0                      16.27906976744186                           5.0  ...   
1                                    0.0                           0.0  ...   
2                                    0.0                           2.5  ...   
3                                    0.0                           0.0  ...   
4                      4.651162790697675                           2.5  ...   
..                                   ...                           ...  ...   
202                   2.3255813953488373                           0.0  ...   
203                    4.651162790697675                           0.0  ...   
204                                  0.0                           0.0  ...   
205                                  0.0                           5.0  ...   
206                                  0.0                           0.0  ...   

    MedianFoldChange_CI_lower MedianFoldChange_CI_upper MedianFoldChange_sig  \
0          0.8062148926982827         1.006288181908508                 True   
1          0.9819002229507531        1.3257967430130182                 True   
2            1.18961009897293         2.049093606776825                 True   
3          0.6131694532285143         1.103634712784355                 True   
4          0.8564560319575159          1.06783254357279                 True   
..                        ...                       ...                  ...   
202                       nan                       nan                False   
203                       nan                       nan                False   
204                       nan                       nan                False   
205                       nan                       nan                False   
206                       nan        

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display edge data used in the Hierarchical Edge Bundle

Check the edge data simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(edges)</span><br>
</div>

In [27]:
display(edges)

start_index start_name           start_label     start_block  end_index  \
0               0         A1  1-Methylnicotinamide  Gastric Cancer          1   
1               0         A1  1-Methylnicotinamide  Gastric Cancer          4   
2               0         A1  1-Methylnicotinamide  Gastric Cancer          6   
3               0         A1  1-Methylnicotinamide  Gastric Cancer          8   
4               0         A1  1-Methylnicotinamide  Gastric Cancer          9   
...           ...        ...                   ...             ...        ...   
5607          201        C64                  u233         Healthy        206   
5608          202        C65                   u43         Healthy        203   
5609          202        C65                   u43         Healthy        206   
5610          204        C67                   u87         Healthy        205   
5611          205        C68     π-Methylhistidine         Healthy        206   

     end_name             end_label       end_block     score  sign    pvalue  
0          A2        2-Aminoadipate  Gastric Cancer  0.509568   1.0  0.000784  
1          A5    3-Aminoisobutyrate  Gastric Cancer  0.397936   1.0  0.010992  
2          A7  3-Hydroxyisovalerate  Gastric Cancer  0.334913   1.0  0.034649  
3          A9      3-Indoxylsulfate  Gastric Cancer  0.447655   1.0  0.003775  
4         A10   6-Hydroxynicotinate  Gastric Cancer  0.427881   1.0  0.005884  
...       ...                   ...             ...       ...   ...       ...  
5607      C69     τ-Methylhistidine         Healthy  0.524953   1.0  0.000506  
5608      C66                  u433         Healthy  0.489585   1.0  0.001345  
5609      C69     τ-Methylhistidine         Healthy  0.504597   1.0  0.000899  
5610      C68     π-Methylhistidine         Healthy  0.316209   1.0  0.046834  
5611      C69     τ-Methylhistidine         Healthy  0.516065   1.0  0.000653  

[5612 rows x 11 columns]

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 11.  Sort edges

Sort the edges for visualisation preference

</div>

In [28]:
#edges.sort_values(['start_index', 'end_index'], inplace=True, ascending=True)
#edges.sort_values('pvalue', inplace=True, ascending=False)
edges.sort_values('score', inplace=True, ascending=False)

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 12.  Plot Hierarchical edge bundle

The edges from the network are then passed into D3 JavaScript to generate a Hierarchical edge bundle and embedded in HTML for interactive visualisation. The Hierarchical edge bundle is implemented as a circular hierarchical tree structure, with nodes on the outside and edges passing through the circle following a bundled curve until they connect to other nodes. The edges represent some association value such as correlation coefficients and can be coloured accordingly based on the values, the sign or represented as pvalues and coloured using a continuous colour map. Different meta data such as groups/classes within the data can also be reflected in the plot to illustrate how the different groups/classes are correlated and to what degree.

Note: The visualisation will automatically open in another tab, unless running in Binder (See step 13).

</div>

In [29]:
bundle = multivis.edgeBundle(nodes,edges)

bundle.help()

Produces an interactive hierarchical edge bundle in D3.js, from nodes and edges.

        Parameters
        ----------
        nodes : Pandas dataframe containing nodes generated from Edge.
        edges : Pandas dataframe containing edges generated from Edge.
        
        Methods
        -------
        set_params : Set parameters -
            html_file: Name to save the HTML file as (default: 'hEdgeBundle.html')
            innerRadiusOffset: Sets the inner radius based on the offset value from the canvas width/diameter (default: 120)
            blockSeparation: Value to set the distance between different segmented blocks (default: 1)
            linkFadeOpacity: The link fade opacity when hovering over/clicking nodes (default: 0.05)
            mouseOver: Setting to 'True' swaps from clicking to hovering over nodes to select them (default: True)
            fontSize: The font size in pixels set for each node (default: 10)
            backgroundColor: Set the background colour

In [30]:
params = dict({'html_file': 'hEdgeBundle_multi-block_gastric_cancer.html'      #HTML file name to save to
               , 'innerRadiusOffset': 120           #The offset from the radius to determine the inner radius of the edge bundle
               , 'blockSeparation': 5               #The degree of separation between blocks
               , 'linkFadeOpacity': 0.01            #The opacity of faded links
               , 'mouseOver': True                  #Setting to 'True' swaps from clicking to hovering over nodes to select them 
               , 'fontSize': 10                     #The font size of each node
               , 'backgroundColor': 'white'         #Set the background colour of the plot
               , 'foregroundColor': 'black'         #Set the foreground colour of the plot
               , 'node_data': ['Name', 'Label', 'VIP1', 'Pvalue', 'MannWhitneyU_pvalue', 'MannWhitneyU_statistic','MannWhitneyU_BHFDR_qvalue', 'Group_median', 'Group_median_sig', 'MedianFoldChange', 'MedianFoldChange_sig', 'PC1', 'PC2', 'PC1_sig', 'PC2_sig']
               , 'nodeColorScale': 'linear'         #The scale to use for colouring the nodes
               , 'node_color_column': 'Pvalue'      #If node_color_column contains colour values it overides the use of node_cmap
               , 'node_cmap': 'brg'                 #Set the colour palette to use for colouring the nodes               
               , 'edge_color_value': 'sign'         #Set the values to colour the edges by. Either 'sign', 'score' or 'pvalue'.
               , 'edgeColorScale': 'linear'         #The scale to use for colouring the edges (if edge_color_value is 'pvalue')
               , 'edge_cmap': 'cool'                #Set the colour palette to use for colouring the edges
               , 'addArcs': True                    #Setting to 'True' adds arcs around the edge bundle for each block
               , 'arcRadiusOffset': 20              #Sets the arc radius offset from the inner radius
               , 'extendArcAngle': 2                #Sets the angle value to add to each end of the arcs 
               , 'arc_cmap': 'tab20'})              #Sets the CMAP colour palette to use for colouring the arcs

bundle.set_params(**params)

bundle.build()

HTML writen to hEdgeBundle_multi-block_gastric_cancer.html


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Build Dashboard

A dashboard with panels for the hierarchical edge bundle, node data and sliders is built, allowing for a more robust interface for exploratory analysis of the data. The dashboard is automatically written to HTML and launched upon creation.

</div>

In [31]:
bundle.buildDashboard()

HTML writen to hEdgeBundle_multi-block_gastric_cancer_dashboard.html


<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 13.  Alternative visualisation options

The visualisations will automatically open when run locally in Jupyter notebook. However, when running in Binder the visualisations will not open in a new tab due to security restrictions within Binder. Therefore, the following options are available: IFrames and JavaScript.

Note 1: Due to security restrictions within Jupyter notebook, resizing the window opened from within Jupyter notebook with Javascript or IFrame will result in a 403: Forbidden error, due to the visualisation being reloaded each time. In this case, just change the dimensions set in the cell and rerun the cell.

Note 2: Depending on the browser you're using the save button may be disabled when opening the visualisation in an IFrame or JavaScript pop-up. In Chrome downloads have been disable to prevent malicious behavior. If you have trouble downloading try a different browser or use the visualisation opened automatically in another tab when running Jupyter notebook locally.
</div>

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Plain visualisation
</div>

In [32]:
vis_option = "IFrame"  #"Javascript"
file = params['html_file']

if vis_option.lower() == "javascript":
    display(Javascript('''window.open(\'{}\','edgeBundle','width=1000,height=1000')'''.format(file)))
elif vis_option.lower() == "iframe":
    display(IFrame(src=file, width='100%', height='1000px'))

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Visualisation with dashboard
</div>

In [33]:
vis_option = "javascript" #"IFrame"
file = params['html_file'].split(".")[0]+"_dashboard.html"

if vis_option.lower() == "javascript":
    display(Javascript('''window.open(\'{}\','edgeBundle','width=1500,height=1000')'''.format(file)))
elif vis_option.lower() == "iframe":
    display(IFrame(src=file, width='100%', height='1000px'))

<IPython.core.display.Javascript object>